File: agent/bot/logger.js
const fs = require('fs')
const path = require('path')

let _currentTaskId = null
let _botId = ''

function setTaskId(id) { _currentTaskId = id || null }

function _sanitizeLabel(value) {
    return String(value || '')
        .trim()
        .replace(/[^A-Za-z0-9_-]+/g, '-')
        .replace(/^[-_]+|[-_]+$/g, '') || 'bot'
}

function _resolveLogLabel() {
    const botId = String(process.env.BOT_ID || '').trim()
    if (botId) return _sanitizeLabel(botId)

    const dataDir = String(process.env.BOT_DATA_DIR || '').trim()
    if (dataDir) {
        const base = path.basename(path.normalize(dataDir))
        if (/^bot\d+$/i.test(base)) return _sanitizeLabel(base)
    }

    const mcUsername = String(process.env.MC_USERNAME || '').trim()
    if (/^Agent\d+$/i.test(mcUsername)) return _sanitizeLabel(mcUsername)
    return ''
}

function _filenameStamp() {
    const now = new Date()
    const pad = (n) => String(n).padStart(2, '0')
    return `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}-${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`
}

function _formatArg(arg) {
    if (typeof arg === 'string') return arg
    if (arg instanceof Error) return arg.stack || arg.message
    try {
        return JSON.stringify(arg)
    } catch (_) {
        return String(arg)
    }
}

function _cleanupOldLogs(logDir, keepDays = 7) {
    const cutoffMs = Date.now() - keepDays * 86400 * 1000
    try {
        for (const fname of fs.readdirSync(logDir)) {
            const fpath = path.join(logDir, fname)
            try {
                if (fs.statSync(fpath).mtimeMs < cutoffMs) fs.unlinkSync(fpath)
            } catch (_) {}
        }
    } catch (_) {}
}

function initLogger(name = 'bot') {
    if (global.__agentLoggerInitialized) return global.__agentLoggerPath

    const logDir = path.join(__dirname, '..', 'logs')
    fs.mkdirSync(logDir, { recursive: true })
    _cleanupOldLogs(logDir)
    const botLabel = _resolveLogLabel()
    _botId = botLabel || 'bot'
    const filename = botLabel ? `${name}-${botLabel}-${_filenameStamp()}.jsonl` : `${name}-${_filenameStamp()}.jsonl`
    const logPath = path.join(logDir, filename)

    const original = {
        log: console.log.bind(console),
        warn: console.warn.bind(console),
        error: console.error.bind(console),
    }

    const write = (level, args) => {
        const entry = {
            time: new Date().toISOString(),
            level,
            service: name,
            bot_id: _botId,
            task_id: _currentTaskId,
            msg: args.map(_formatArg).join(' '),
        }
        fs.appendFileSync(logPath, JSON.stringify(entry) + '\n', 'utf8')
    }

    console.log = (...args) => {
        write('INFO', args)
        original.log(...args)
    }
    console.warn = (...args) => {
        write('WARN', args)
        original.warn(...args)
    }
    console.error = (...args) => {
        write('ERROR', args)
        original.error(...args)
    }

    global.__agentLoggerInitialized = true
    global.__agentLoggerPath = logPath
    console.log(`[Log] 已寫入 ${logPath}`)
    return logPath
}

module.exports = { initLogger, setTaskId }

File: agent/logger.py
from __future__ import annotations

import json
import os
import re
import sys
from datetime import datetime, timezone

_current_task_id: str | None = None


def set_task_id(tid: str | None) -> None:
    global _current_task_id
    _current_task_id = tid


def get_task_id() -> str | None:
    return _current_task_id


class _TeeStream:
    def __init__(self, original, file_handle, level: str, service: str, bot_id: str):
        self._original = original
        self._file_handle = file_handle
        self._level = level
        self._service = service
        self._bot_id = bot_id

    def write(self, data: str) -> int:
        written = self._original.write(data)
        if data and not data.isspace():
            for line in data.splitlines(True):
                if not line.strip():
                    self._file_handle.write("\n")
                    continue
                msg = line.rstrip("\n")
                # [State] tick lines are terminal-only — too noisy for log file
                if msg.lstrip().startswith("[State]"):
                    continue
                entry = {
                    "time": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
                    "level": self._level,
                    "service": self._service,
                    "bot_id": self._bot_id,
                    "task_id": _current_task_id,
                    "msg": msg,
                }
                self._file_handle.write(json.dumps(entry, ensure_ascii=False) + "\n")
                try:
                    import agent.history_db as _hdb
                    _hdb._fire_and_forget(_hdb.write_log, entry)
                except Exception:
                    pass
        self._file_handle.flush()
        return written

    def flush(self) -> None:
        self._original.flush()
        self._file_handle.flush()

    def isatty(self) -> bool:
        return getattr(self._original, "isatty", lambda: False)()


def init_logger(service: str = "agent") -> str:
    if getattr(sys, "_agent_log_initialized", False):
        return getattr(sys, "_agent_log_path")

    bot_id = _resolve_bot_id()
    base_dir = os.path.dirname(__file__)
    log_dir = os.path.join(base_dir, "logs")
    os.makedirs(log_dir, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    bot_label = _resolve_log_label()
    filename = f"{service}-{bot_label}-{stamp}.jsonl" if bot_label else f"{service}-{stamp}.jsonl"
    log_path = os.path.join(log_dir, filename)
    file_handle = open(log_path, "a", encoding="utf-8", buffering=1)

    sys.stdout = _TeeStream(sys.stdout, file_handle, "INFO", service, bot_id)
    sys.stderr = _TeeStream(sys.stderr, file_handle, "ERROR", service, bot_id)
    sys._agent_log_initialized = True
    sys._agent_log_path = log_path
    print(f"[Log] 已寫入 {log_path}")
    return log_path


def _resolve_bot_id() -> str:
    return (os.getenv("BOT_ID") or "").strip() or "bot0"


def _resolve_log_label() -> str | None:
    bot_id = (os.getenv("BOT_ID") or "").strip()
    if bot_id:
        return _sanitize_label(bot_id)

    data_dir = (os.getenv("BOT_DATA_DIR") or "").strip()
    if data_dir:
        base = os.path.basename(os.path.normpath(data_dir))
        if re.fullmatch(r"bot\d+", base, flags=re.IGNORECASE):
            return _sanitize_label(base)

    mc_username = (os.getenv("MC_USERNAME") or "").strip()
    if re.fullmatch(r"Agent\d+", mc_username, flags=re.IGNORECASE):
        return _sanitize_label(mc_username)
    return None


def _sanitize_label(value: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9_-]+", "-", value.strip())
    return cleaned.strip("-_") or "bot"
